File: 필수문항/147_비슷한단어.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        String basic = br.readLine();
        int ans = 0;
        while (N-- > 1) {
            char[] word = br.readLine().toCharArray();
            int onlyBasic = 0;
            int onlyWord = 0;
            for (int i = 0; i < basic.length(); i++) {
                boolean check = false;
                for (int j = 0; j < word.length; j++) {
                    if (word[j] == basic.charAt(i)) {
                        word[j] = '*';
                        check = true;
                        break;
                    }
                }
                if (!check) {
                    onlyBasic++;
                }
            }
            for (char c : word) {
                if (c != '*') {
                    onlyWord++;
                }
            }
            if (Math.abs(basic.length() - word.length) <= 1 && onlyWord <= 1 && onlyBasic <= 1) {
                ans++;
            }

        }
        System.out.println(ans);
    }
}

File: 필수문항/147_비슷한단어.py
import sys
from collections import defaultdict

input = sys.stdin.readline
n = int(input())
str_basic = input().rstrip()
lb = len(str_basic)
basic = defaultdict(int)
for c in str_basic:
    basic[c] += 1

ans = 0


def same(word):
    global basic
    for key, value in basic.items():
        word[key] -= value

    check_plus = False
    check_minus = False

    for value in word.values():
        if abs(value) > 1:
            return False
        if value == 1:
            if check_plus:
                return False
            check_plus = True
        if value == -1:
            if check_minus:
                return False
            check_minus = True
    return True


def lw_one(basic_copy, word):
    check = 0
    for key, value in word.items():
        check += abs(value - basic_copy[key])
        if check > 1:
            return False
    return True


def lb_one(word):
    check = 0
    for key, value in basic.items():
        check += abs(value - word[key])
        if check > 1:
            return False
    return True


for _ in range(n - 1):
    str_word = input().rstrip()
    lw = len(str_word)
    word = defaultdict(int)
    for c in str_word:
        word[c] += 1

    if lb == lw:
        if same(word):
            ans += 1

    elif lb + 1 == lw:
        basic_copy = basic.copy()
        if lw_one(basic_copy, word):
            ans += 1

    elif lb == lw + 1:
        if lb_one(word):
            ans += 1


print(ans)
